stl/inc/spanstream

```diff
@@ -0,0 +1,386 @@
+        }
+
+        if (_Sequence_out) {
+            _Mysb::pbump(static_cast<int>(_Off - (_Mysb::pptr() - _Mysb::pbase())));
+        }
+
+        return pos_type(_Off);
+    }
+
+    pos_type seekpos(pos_type _Pos, ios_base::openmode _Which = ios_base::in | ios_base::out) override {
+        return seekoff(off_type{_Pos}, ios_base::beg, _Which);
+    }
+
+    _Mysb* setbuf(_Elem* _Buffer, streamsize _Count) override {
+        this->span(_STD span<_Elem>{_Buffer, static_cast<size_t>(_Count)});
+        return this;
+    }
+
+private:
+    ios_base::openmode _Mode{ios_base::in | ios_base::out};
+    _STD span<_Elem> _Buf{};
+};
+
+template <class _Elem, class _Traits>
+void swap(basic_spanbuf<_Elem, _Traits>& _Left, basic_spanbuf<_Elem, _Traits>& _Right) {
+    _Left.swap(_Right);
+}
+
+template <class _Elem, class _Traits>
+class basic_ispanstream : public basic_istream<_Elem, _Traits> {
+public:
+    using _Mybase     = basic_istream<_Elem, _Traits>;
+    using _Mysb       = basic_spanbuf<_Elem, _Traits>;
+    using char_type   = _Elem;
+    using int_type    = typename _Traits::int_type;
+    using pos_type    = typename _Traits::pos_type;
+    using off_type    = typename _Traits::off_type;
+    using traits_type = _Traits;
+
+    // N4892 [ispanstream.ctor], constructors
+    explicit basic_ispanstream(_STD span<_Elem> _Span, ios_base::openmode _Which = ios_base::in)
+        : _Mybase(_STD addressof(_Buf)), _Buf(_Span, _Which | ios_base::in) {}
+
+    basic_ispanstream(const basic_ispanstream&) = delete;
+
+    basic_ispanstream(basic_ispanstream&& _Right) : _Mybase(_STD move(_Right)), _Buf(_STD move(_Right._Buf)) {
+        _Mybase::set_rdbuf(_STD addressof(_Buf));
+    }
+
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    template <_RANGES borrowed_range _ReadOnlyRange>
+        requires (!convertible_to<_ReadOnlyRange, _STD span<_Elem>>
+                && convertible_to<_ReadOnlyRange, _STD span<const _Elem>>)
+    explicit basic_ispanstream(_ReadOnlyRange&& _Range)
+        : basic_ispanstream(
+            _STD span<_Elem>{const_cast<_Elem*>(_RANGES data(_Range)), static_cast<size_t>(_RANGES size(_Range))}) {}
+    // clang-format on
+#endif // __cpp_lib_concepts
+
+    // N4892 [ispanstream.assign], assignment and swap
+    basic_ispanstream& operator=(const basic_ispanstream&) = delete;
+
+    basic_ispanstream& operator=(basic_ispanstream&& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    void swap(basic_ispanstream& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    // N4892 [ispanstream.members], member functions
+    _Mysb* rdbuf() const noexcept {
+        return const_cast<_Mysb*>(_STD addressof(_Buf));
+    }
+
+    span<const _Elem> span() const noexcept {
+        return this->rdbuf()->span();
+    }
+
+    void span(_STD span<_Elem> _Span) noexcept {
+        this->rdbuf()->span(_Span);

```

Comment:
`rdbuf()` is a direct member function, so we don't need to say `this->`.

Occurs below.